File: zig/src/buffer.zig
//! Gap buffer implementation
//! Text storage with cursor always positioned at the gap

const std = @import("std");

pub const GapBuffer = struct {
    data: []u8,
    gap_start: usize,
    gap_end: usize,
    allocator: std.mem.Allocator,
    initialized: bool = true, // Track if buffer was properly initialized

    const INITIAL_CAPACITY = 1024;
    const MIN_GAP_SIZE = 16;

    pub fn init(allocator: std.mem.Allocator) !GapBuffer {
        const capacity = INITIAL_CAPACITY;
        const data = try allocator.alloc(u8, capacity);
        const gap_start: usize = 0;
        const gap_end: usize = capacity;

        return GapBuffer{
            .data = data,
            .gap_start = gap_start,
            .gap_end = gap_end,
            .allocator = allocator,
            .initialized = true,
        };
    }

    pub fn deinit(self: *GapBuffer) void {
        // Free the buffer data only if it was properly initialized
        // Check both initialized flag and that we have a valid allocation
        if (self.initialized and self.data.len > 0) {
            self.allocator.free(self.data);
        }
        self.* = undefined;
    }

    /// Insert bytes at the current cursor position (gap position)
    pub fn insert(self: *GapBuffer, bytes: []const u8) !void {
        if (bytes.len == 0) return;

        // Ensure gap is large enough
        const gap_size = self.gap_end - self.gap_start;
        if (gap_size < bytes.len) {
            try self.expandGap(bytes.len);
        }

        // Copy bytes into gap
        @memcpy(self.data[self.gap_start..self.gap_start + bytes.len], bytes);
        self.gap_start += bytes.len;
    }

    /// Delete the byte before the cursor (before gap)
    pub fn delete(self: *GapBuffer) bool {
        if (self.gap_start == 0) return false;
        self.gap_start -= 1;
        return true;
    }

    /// Move cursor left (move gap left)
    pub fn moveLeft(self: *GapBuffer) bool {
        if (self.gap_start == 0) return false;
        self.gap_start -= 1;
        self.gap_end -= 1;
        self.data[self.gap_end] = self.data[self.gap_start];
        return true;
    }

    /// Move cursor right (move gap right)
    pub fn moveRight(self: *GapBuffer) bool {
        if (self.gap_end >= self.data.len) return false;
        self.data[self.gap_start] = self.data[self.gap_end];
        self.gap_start += 1;
        self.gap_end += 1;
        return true;
    }

    /// Move cursor up one line
    pub fn moveUp(self: *GapBuffer) bool {
        const line_start = self.getLineStart(self.gap_start);
        if (line_start == 0) return false; // Already at first line

        // Find start of previous line
        var prev_line_start = line_start - 1;
        while (prev_line_start > 0 and self.data[prev_line_start - 1] != '\n') {
            prev_line_start -= 1;
        }

        // Calculate column position in current line
        const current_col = self.gap_start - line_start;

        // Find target position in previous line
        var target_pos = prev_line_start;
        var col_count: usize = 0;
        while (target_pos < line_start and col_count < current_col) {
            if (self.data[target_pos] == '\n') break;
            target_pos += 1;
            col_count += 1;
        }

        // Move gap to target position
        return self.moveGapTo(target_pos);
    }

    /// Move cursor down one line
    pub fn moveDown(self: *GapBuffer) bool {
        const line_end = self.getLineEnd(self.gap_start);
        const content_end = self.getContentEnd();
        if (line_end >= content_end) return false; // Already at last line

        // Find start of next line
        const next_line_start = line_end + 1;
        if (next_line_start >= self.getContentEnd()) return false;

        // Calculate column position in current line
        const line_start = self.getLineStart(self.gap_start);
        const current_col = self.gap_start - line_start;

        // Find target position in next line
        var target_pos = next_line_start;
        var col_count: usize = 0;
        const next_line_end = self.getLineEnd(next_line_start);
        while (target_pos < next_line_end and col_count < current_col) {
            if (self.data[target_pos] == '\n') break;
            target_pos += 1;
            col_count += 1;
        }

        // Move gap to target position
        return self.moveGapTo(target_pos);
    }

    /// Move to start of buffer
    pub fn moveToStart(self: *GapBuffer) void {
        _ = self.moveGapTo(0);
    }

    /// Move to end of buffer
    pub fn moveToEnd(self: *GapBuffer) void {
        const end = self.getContentEnd();
        _ = self.moveGapTo(end);
    }

    /// Move to start of current line
    pub fn moveToLineStart(self: *GapBuffer) void {
        const line_start = self.getLineStart(self.gap_start);
        _ = self.moveGapTo(line_start);
    }

    /// Move to end of current line
    pub fn moveToLineEnd(self: *GapBuffer) void {
        const line_end = self.getLineEnd(self.gap_start);
        _ = self.moveGapTo(line_end);
    }

    /// Get line number at cursor (0-indexed)
    pub fn getLine(self: *const GapBuffer) usize {
        var line: usize = 0;
        var pos: usize = 0;
        const content_end = self.getContentEnd();

        while (pos < self.gap_start and pos < content_end) {
            if (self.data[pos] == '\n') {
                line += 1;
            }
            pos += 1;
        }

        return line;
    }

    /// Get total number of lines
    pub fn getTotalLines(self: *const GapBuffer) usize {
        var lines: usize = 1; // At least one line
        const content_end = self.getContentEnd();

        for (0..content_end) |i| {
            if (self.data[i] == '\n') {
                lines += 1;
            }
        }

        return lines;
    }

    /// Get start position of line containing pos
    fn getLineStart(self: *const GapBuffer, pos: usize) usize {
        var start = pos;

        // Adjust for gap
        if (start > self.gap_start) {
            start += (self.gap_end - self.gap_start);
        }

        // Find line start
        while (start > 0 and self.data[start - 1] != '\n') {
            start -= 1;
        }

        // Adjust back for gap
        if (start > self.gap_start) {
            start -= (self.gap_end - self.gap_start);
        }

        return start;
    }

    /// Get end position of line containing pos (position of newline or end of buffer)
    fn getLineEnd(self: *const GapBuffer, pos: usize) usize {
        var end = pos;
        const content_end = self.getContentEnd();

        // Adjust for gap
        if (end > self.gap_start) {
            end += (self.gap_end - self.gap_start);
        }

        // Find line end
        while (end < content_end and self.data[end] != '\n') {
            end += 1;
        }

        // Adjust back for gap
        if (end > self.gap_start) {
            end -= (self.gap_end - self.gap_start);
        }

        return end;
    }

    /// Get the end of actual content (excluding gap)
    fn getContentEnd(self: *const GapBuffer) usize {
        return self.data.len - (self.gap_end - self.gap_start);
    }

    /// Move gap to a specific position
    fn moveGapTo(self: *GapBuffer, target_pos: usize) bool {
        const content_end = self.getContentEnd();
        if (target_pos > content_end) return false;

        const current_pos = self.gap_start;

        if (target_pos == current_pos) return true;

        if (target_pos < current_pos) {
            // Move gap left: move bytes from before gap to after gap
            const bytes_to_move = current_pos - target_pos;
            var i: usize = 0;
            while (i < bytes_to_move) {
                self.gap_end -= 1;
                self.gap_start -= 1;
                self.data[self.gap_end] = self.data[self.gap_start];
                i += 1;
            }
        } else {
            // Move gap right: move bytes from after gap to before gap
            const bytes_to_move = target_pos - current_pos;
            var i: usize = 0;
            while (i < bytes_to_move) {
                self.data[self.gap_start] = self.data[self.gap_end];
                self.gap_start += 1;
                self.gap_end += 1;
                i += 1;
            }
        }

        return true;
    }

    /// Expand gap to accommodate at least min_size bytes
    fn expandGap(self: *GapBuffer, min_size: usize) !void {
        const current_gap_size = self.gap_end - self.gap_start;
        const needed_size = min_size + MIN_GAP_SIZE;
        const new_capacity = self.data.len + (needed_size - current_gap_size);

        const old_data = self.data;
        const new_data = try self.allocator.alloc(u8, new_capacity);

        // Copy data before gap
        @memcpy(new_data[0..self.gap_start], old_data[0..self.gap_start]);

        // Copy data after gap
        const after_gap_start = self.gap_end;
        const after_gap_len = old_data.len - after_gap_start;
        const new_gap_end = self.gap_start + needed_size;
        @memcpy(new_data[new_gap_end..new_gap_end + after_gap_len], old_data[after_gap_start..]);

        // Update buffer state only after successful allocation and copy
        self.data = new_data;
        self.gap_end = new_gap_end;
        
        // Free old data only after everything succeeded
        self.allocator.free(old_data);
    }

    /// Get all content as a slice (for rendering)
    /// Returns a slice that represents the content without the gap
    pub fn getContent(self: *const GapBuffer) []const u8 {
        const before_gap = self.data[0..self.gap_start];
        // We can't return a contiguous slice, so callers need to handle this differently
        // For now, return before_gap only - rendering will need special handling
        _ = self.data[self.gap_end..]; // after_gap exists but not returned
        return before_gap;
    }

    /// Get content before gap
    pub fn getBeforeGap(self: *const GapBuffer) []const u8 {
        return self.data[0..self.gap_start];
    }

    /// Get content after gap
    pub fn getAfterGap(self: *const GapBuffer) []const u8 {
        return self.data[self.gap_end..];
    }

    /// Load file content into buffer
    pub fn loadFromFile(self: *GapBuffer, file_path: []const u8) !void {
        const file = try std.fs.cwd().openFile(file_path, .{});
        defer file.close();

        const file_size = try file.getEndPos();
        const content = try file.readToEndAlloc(self.allocator, file_size);
        defer self.allocator.free(content);

        // Clear buffer first
        self.gap_start = 0;
        self.gap_end = self.data.len;

        // Ensure capacity - if we need to reallocate, do it safely
        if (self.data.len < content.len) {
            const old_data = self.data;
            const new_capacity = if (content.len * 2 > INITIAL_CAPACITY) content.len * 2 else INITIAL_CAPACITY;
            self.data = try self.allocator.alloc(u8, new_capacity);
            self.gap_end = new_capacity;
            // Free old data only after successful allocation
            self.allocator.free(old_data);
        }

        // Insert content
        try self.insert(content);
    }

    /// Delete current line (dd command)
    pub fn deleteLine(self: *GapBuffer) bool {
        const line_start = self.getLineStart(self.gap_start);
        const line_end = self.getLineEnd(self.gap_start);

        if (line_start == line_end and line_start >= self.getContentEnd()) {
            return false; // Empty buffer or last empty line
        }

        // Move gap to line start
        _ = self.moveGapTo(line_start);

        // Delete until line end (including newline if present)
        var deleted: usize = 0;
        const target_end = if (line_end < self.getContentEnd() and self.data[line_end] == '\n') line_end + 1 else line_end;

        while (self.gap_end < self.data.len and deleted < (target_end - line_start)) {
            self.gap_end += 1;
            deleted += 1;
        }

        return true;
    }
};

File: src/buffer.rs
//! Gap buffer implementation for efficient text editing

use std::alloc::{alloc, dealloc, Layout};

/// Gap buffer for efficient insertion and deletion
pub struct GapBuffer {
    /// Buffer containing text before gap, gap, and text after gap
    /// Layout: [before_gap][gap][after_gap]
    buffer: *mut u8,
    /// Capacity of the buffer
    capacity: usize,
    /// Start of gap (end of before_gap)
    gap_start: usize,
    /// End of gap (start of after_gap)
    gap_end: usize,
}

impl GapBuffer {
    /// Create a new gap buffer with initial capacity
    pub fn new(initial_capacity: usize) -> Result<Self, String> {
        if initial_capacity == 0 {
            return Err("Capacity must be > 0".to_string());
        }

        let layout = Layout::from_size_align(initial_capacity, 1)
            .map_err(|e| format!("Invalid layout: {}", e))?;
        
        let buffer = unsafe { alloc(layout) };
        if buffer.is_null() {
            return Err("Failed to allocate buffer".to_string());
        }

        Ok(GapBuffer {
            buffer,
            capacity: initial_capacity,
            gap_start: 0,
            gap_end: initial_capacity,
        })
    }

    /// Get the current cursor position (same as gap_start)
    pub fn cursor(&self) -> usize {
        self.gap_start
    }

    /// Get the total length of text (excluding gap)
    pub fn len(&self) -> usize {
        self.gap_start + (self.capacity - self.gap_end)
    }

    /// Check if buffer is empty
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Move cursor left (move gap right)
    pub fn move_left(&mut self) -> bool {
        if self.gap_start > 0 {
            unsafe {
                let byte = *self.buffer.add(self.gap_start - 1);
                *self.buffer.add(self.gap_end - 1) = byte;
            }
            self.gap_start -= 1;
            self.gap_end -= 1;
            true
        } else {
            false
        }
    }

    /// Move cursor right (move gap left)
    pub fn move_right(&mut self) -> bool {
        if self.gap_end < self.capacity {
            unsafe {
                let byte = *self.buffer.add(self.gap_end);
                *self.buffer.add(self.gap_start) = byte;
            }
            self.gap_start += 1;
            self.gap_end += 1;
            true
        } else {
            false
        }
    }

    /// Insert a byte at the cursor position
    pub fn insert(&mut self, byte: u8) -> Result<(), String> {
        if self.gap_start >= self.gap_end {
            // Gap is exhausted, need to grow
            self.grow()?;
        }

        unsafe {
            *self.buffer.add(self.gap_start) = byte;
        }
        self.gap_start += 1;
        Ok(())
    }

    /// Insert a string at the cursor position
    pub fn insert_str(&mut self, s: &str) -> Result<(), String> {
        for byte in s.bytes() {
            self.insert(byte)?;
        }
        Ok(())
    }

    /// Delete the byte before the cursor (backspace)
    pub fn delete_backward(&mut self) -> bool {
        if self.gap_start > 0 {
            // Move gap left by one
            unsafe {
                let byte = *self.buffer.add(self.gap_start - 1);
                *self.buffer.add(self.gap_end - 1) = byte;
            }
            self.gap_start -= 1;
            self.gap_end -= 1;
            true
        } else {
            false
        }
    }

    /// Delete the byte at the cursor position (delete)
    pub fn delete_forward(&mut self) -> bool {
        if self.gap_end < self.capacity {
            // Just expand gap by one (effectively deleting the byte)
            self.gap_end += 1;
            true
        } else {
            false
        }
    }

    /// Get the text before the gap as a string
    pub fn get_before_gap(&self) -> &[u8] {
        unsafe {
            std::slice::from_raw_parts(self.buffer, self.gap_start)
        }
    }

    /// Get the text after the gap as a string
    pub fn get_after_gap(&self) -> &[u8] {
        unsafe {
            std::slice::from_raw_parts(self.buffer.add(self.gap_end), self.capacity - self.gap_end)
        }
    }

    /// Get the entire text as a string (reconstructs by moving gap to end)
    pub fn to_string(&self) -> String {
        let before = self.get_before_gap();
        let after = self.get_after_gap();
        let mut result = Vec::with_capacity(before.len() + after.len());
        result.extend_from_slice(before);
        result.extend_from_slice(after);
        String::from_utf8_lossy(&result).to_string()
    }

    /// Get the line number at the cursor position
    pub fn get_line(&self) -> usize {
        let before = self.get_before_gap();
        before.iter().filter(|&&b| b == b'\n').count()
    }

    /// Get the total number of lines
    pub fn get_total_lines(&self) -> usize {
        let before = self.get_before_gap();
        let after = self.get_after_gap();
        let newlines = before.iter().filter(|&&b| b == b'\n').count()
            + after.iter().filter(|&&b| b == b'\n').count();
        // If there's at least one newline, lines = newlines + 1
        // If no newlines, lines = 1 (single line)
        if newlines > 0 || !before.is_empty() || !after.is_empty() {
            newlines + 1
        } else {
            0
        }
    }

    /// Grow the buffer when gap is exhausted
    fn grow(&mut self) -> Result<(), String> {
        let new_capacity = self.capacity * 2;
        let new_layout = Layout::from_size_align(new_capacity, 1)
            .map_err(|e| format!("Invalid layout: {}", e))?;

        let new_buffer = unsafe { alloc(new_layout) };
        if new_buffer.is_null() {
            return Err("Failed to allocate new buffer".to_string());
        }

        // Copy before_gap
        unsafe {
            std::ptr::copy_nonoverlapping(
                self.buffer,
                new_buffer,
                self.gap_start,
            );
        }

        // Copy after_gap to the end
        let after_len = self.capacity - self.gap_end;
        unsafe {
            std::ptr::copy_nonoverlapping(
                self.buffer.add(self.gap_end),
                new_buffer.add(new_capacity - after_len),
                after_len,
            );
        }

        // Deallocate old buffer
        let old_layout = Layout::from_size_align(self.capacity, 1).unwrap();
        unsafe {
            dealloc(self.buffer, old_layout);
        }

        self.gap_end = new_capacity - after_len;
        self.capacity = new_capacity;
        self.buffer = new_buffer;

        Ok(())
    }
}

impl Drop for GapBuffer {
    fn drop(&mut self) {
        let layout = Layout::from_size_align(self.capacity, 1).unwrap();
        unsafe {
            dealloc(self.buffer, layout);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new() {
        let buf = GapBuffer::new(10).unwrap();
        assert_eq!(buf.len(), 0);
        assert_eq!(buf.cursor(), 0);
    }

    #[test]
    fn test_insert() {
        let mut buf = GapBuffer::new(10).unwrap();
        buf.insert(b'a').unwrap();
        assert_eq!(buf.len(), 1);
        assert_eq!(buf.cursor(), 1);
    }

    #[test]
    fn test_move_and_insert() {
        let mut buf = GapBuffer::new(10).unwrap();
        buf.insert_str("hello").unwrap();
        // Move to start
        for _ in 0..5 {
            buf.move_left();
        }
        buf.insert(b'X').unwrap();
        assert_eq!(buf.to_string(), "Xhello");
    }

    #[test]
    fn test_delete() {
        let mut buf = GapBuffer::new(10).unwrap();
        buf.insert_str("hello").unwrap();
        buf.move_left();
        buf.delete_backward().unwrap();
        assert_eq!(buf.to_string(), "hell");
    }
}

